test/route.test.js

```diff
@@ -217,6 +217,59 @@ test('same route definition object on multiple prefixes', async t => {
   await fastify.ready()
 })
 
+test('mutating the route definition in onRoute hooks shouldnt interfere with afterRouteAdded() being called twice', t => {
+  // Since afterRouteAdded is called twice in this test, we have to plan 2 asserts for each assert statement inside a hook
+  t.plan(12)
+
+  const routeObject = {
+    handler: (req, reply) => { reply.send('hi from handler') },
+    method: 'POST',
+    url: '/'
+  }
+
+  const fastify = Fastify()
+  t.tearDown(() => fastify.close())
+
+  let onRequestCallCounter = 0
+
+  fastify.register(async function (f) {
+    f.addHook('onRoute', (routeOptions) => {
+      // We make sure that the second time afterRouteAdded() is called, we receive a new copy of the original routeObject
+      t.is(routeOptions.method, 'POST')
+      t.is(routeOptions.onRequest, undefined)
+
+      routeOptions.method = 'GET'
+      routeOptions.onRequest = [onRequest]
+
+      function onRequest (req, reply, next) {
+        onRequestCallCounter++
+        next()
+      }
+    })
+    f.addHook('onRoute', (routeOptions) => {
+      // We make sure that this hook receives the object mutated by the previous onRoute hook executed
+      t.is(routeOptions.method, 'GET')
+      t.is(routeOptions.onRequest.length, 1)
+    })
+
+    f.route(routeObject)
+  }, { prefix: '/v1' })
+
+  fastify.listen(0, err => {
+    t.error(err)
+    sget({
+      method: 'GET',
+      url: 'http://localhost:' + fastify.server.address().port + '/v1/'
+    }, (err, response, body) => {
+      t.error(err)
+      // The original object shouldn't be mutated
+      t.is(routeObject.method, 'POST')
+      // The onRequest hook was added & called once
+      t.equal(onRequestCallCounter, 1)
+    })
+  })
```

Comment:
This is not mapping 1:1 to the compress case. Could you add _another_ onRequest hook (not replacing it).

Reply:
My understanding of that issue is that for `fastify-compress` the `onRoute` is called twice and the `onSend` hook is added twice to the route definition. This results in doubly-compressed response data.

My fix here bypasses that and ensures that the the `onSend` hook is only added once.
Anyway, I will update the test to add another onRequest hook.

Also, the issue I mentioned in Salman's PR happens here too in some cases, will provide a fix for it as well.

Reply:
I mean that you should test if an `onRoute` hook is already present and `onRoute` receives an array.

Reply:
> I mean that you should test if an `onRoute` hook is already present

Do you mean test if an `onRequest` hook is already present? If not, how would I check if an `onRoute` hook is already present? AFAIK the route definition doesn't contain `onRoute` hooks.

> and `onRoute` receives an array.

The test I added does show that it received an `onRequest` array.

Here's what happens. The test contains two `onRoute` hooks:
- The first one makes sure that `onRequest` is `undefined` then sets it.
- The second makes sure that it receives an `onRequest` array with one element (set by the previous hook)

This displays that:
- `onRoute` hooks are a reduce operation where they can all mutate the route options and only the last is registered
- route options mutated by `/v1` `onRoute` hooks are not passed to `/v1/` hooks

With that said, I like the solution provided by salman more because it makes sense from an end user perspective that `onRoute` hooks would be called only once for `/v1` and reuse the mutated route options for `/v1/`, so feel free to close this PR if you agree.

Reply:
> The test I added does show that it received an onRequest array.

FYI, the problem happens when the array is mutated.

> I like the solution provided by salman more because it makes sense from an end user perspective that onRoute hooks would be called only once for /v1 and reuse the mutated route options for /v1/, so feel free to close this PR if you agree.

i'll close this PR then, thanks!